### PR TITLE
[MRG+1]: Unify Raw and Epochs-based covariance

### LINKED ISSUE #2867: compute_raw_covariance vs compute_covariance
Looking at `compute_raw_covariance`, it is not as good as `compute_covariance` because it can't make use of the advanced regularization methods. Looking at the code, it seems like we could have `compute_raw_covariance` create an `Epochs` instance under the hood using `make_fixed_length_events` and then call `compute_covariance`. @dengemann I think we discussed doing this before, is there some reason we decided not to do it that way?

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #2867. Related to #2448.

This simplifies our code a bit, and also shows how `compute_raw_covariance` differed from `compute_covariance` under the hood.

As shown in the code, `compute_raw_covariance` has (effectively) `keep_sample_mean=False` by default, whereas `compute_covariance` has `keep_sample_mean=True` by default. This makes me think we should have `keep_sample_mean=None` -> `False` with deprecation message saying it will change to `True` in 0.13. This would really unify the two functions. Thoughts?

### CODE REVIEW

File: mne/tests/test_cov.py [hunk 1]
```diff
@@ -101,9 +101,9 @@ def test_cov_estimation_on_raw_segment():
     raw = Raw(raw_fname, preload=False)
     cov = compute_raw_covariance(raw)
     cov_mne = read_cov(erm_cov_fname)
-    assert_true(cov_mne.ch_names == cov.ch_names)
-    assert_true(linalg.norm(cov.data - cov_mne.data, ord='fro') /
-                linalg.norm(cov.data, ord='fro') < 1e-4)
+    assert_equal(cov_mne.ch_names, cov.ch_names)
+    assert_equal(cov_mne.nfree, cov.nfree)
+    assert_snr(cov.data, cov_mne.data, 100)
```
Comment (Eric89GXL) [inline mne/tests/test_cov.py]: I had to lower the tolerance here from 10000 to 100, but by eye looking at `cov.plot()` comparisons of spectra and images of the covariances they look very similar, so I'm pretty sure the calculations are nearly identical.

File: mne/cov.py [hunk 2]
```diff
@@ -386,66 +389,90 @@ def compute_raw_covariance(raw, tmin=None, tmax=None, tstep=0.2,
...
-    mu = 0
-
-    info = pick_info(raw.info, picks)
-    idx_by_type = channel_indices_by_type(info)
-
-    # Read data in chunks
-    for first in range(start, stop, step):
-        last = first + step
-        if last >= stop:
-            last = stop
```
Comment (Eric89GXL) [inline mne/cov.py]: This change could fall into the bugfix category. This used to allow an epoch of shorter length than all the others to be included in the calculations, the new code does not (`Epochs` constructor will throw out such segments).

File: mne/cov.py [hunk 3]
```diff
@@ -386,66 +389,90 @@ def compute_raw_covariance(raw, tmin=None, tmax=None, tstep=0.2,
...
+        previous stimulus extends into the baseline period of the next.
+        Note. This option is only implemented for method='empirical'.
+        If True, the epochs will only be baseline corrected.
+
+        .. versionadded:: 0.12
+
+    method : str | list | None (default 'empirical')
+        The method used for covariance estimation.
+        See :func:`mne.compute_covariance`.
+
```
Comment (dengemann) [inline mne/cov.py]: maybe add a note that  'auto' might not be the best choice here as it can be very time-consuming to compute on full raw data.

**Comment by dengemann** [comment]
Ok your recent arguments convinced me. Seems like a good idea.

**Comment by Eric89GXL** [comment]
@dengemann any opinion on the `keep_sample_mean` differences? I haven't thought too much about this parameter so I'm not sure what the best thing to do is.

**Comment by Eric89GXL** [comment]
@agramfort any opinion on what to do with `keep_sample_mean`? It is `True` for `compute_covariance` but effectively `False` (was hidden but now exposed) for `compute_raw_covariance`. Should we unify them through deprecating `False` for `compute_raw_covariance`?

**Comment by dengemann** [comment]
@Eric89GXL I'm wondering which meaning it can have here, as we don't have different conditions in continues mode.

**Comment by Eric89GXL** [comment]
It changes the results a bit depending on whether or not it's enabled

File: mne/cov.py [hunk 4]
```diff
@@ -386,66 +389,90 @@ def compute_raw_covariance(raw, tmin=None, tmax=None, tstep=0.2,
     picks : array-like of int | None (default None)
         Indices of channels to include (if None, all channels
         except bad channels are used).
+    keep_sample_mean : bool (default False)
```
Comment (agramfort) [inline mne/cov.py]: there is no event here. Docstring does not make sense for raw.

Comment (Eric89GXL) [inline mne/cov.py]: So instead it should probably read "the average response over epochs is computed and subtracted during the covariance computation"

Comment (dengemann) [inline mne/cov.py]: Yes, exactly. I think somehow this option should even be deactivated in
continous mode ...

Comment (Eric89GXL) [inline mne/cov.py]: But it was `False` by default before. So you agree there should be a deprecation cycle where we make it `True` by default? Or remove it entirely, disabling the old behavior altogether?

Comment (agramfort) [inline mne/cov.py]: what do you mean it was False before? it was not there so no deprecation and let's keep the behavior unchanged. ok?

Comment (dengemann) [inline mne/cov.py]: I would just not expose it here, it does not make sense for continuous recordings.

Comment (agramfort) [inline mne/cov.py]: +1

Comment (Eric89GXL) [inline mne/cov.py]: If the parameter is not exposed, then that means making it `keep_sample_mean=True` or `keep_sample_mean=False` always.

Always using `keep_sample_mean=True` will make getting the old behavior of `compute_raw_covariance`, and thus the behavior equivalent to the version computed with `mne_process_raw` in the sample dataset, impossible to get in mne-python.

Always using `keep_sample_mean=False` will make it impossible to match the behavior of `compute_covariance` if you just did `compute_covariance(Epochs(..., make_fixed_length_events(...), ...))` with no `keep_sample_mean` arg (because there it's default `True`).

Which one are you supporting?

Comment (Eric89GXL) [inline mne/cov.py]: > what do you mean it was False before?

This must be the source of confusion. The way `compute_raw_covariance` works in `master` is equivalent to using `compute_raw_covariance(..., keep_sample_mean=False)` in the current PR. So proposing to have no `keep_sample_mean` arg and have it behave like `keep_sample_mean=True` means you can no longer get a computation equivalent to current `master` for `compute_raw_covariance`. Make sense?

Comment (Eric89GXL) [inline mne/cov.py]: Or to put it differently, as in the original description:

> As shown in the code, compute_raw_covariance has (effectively) keep_sample_mean=False by default, whereas compute_covariance has keep_sample_mean=True by default. This makes me think we should have keep_sample_mean=None -> False with deprecation message saying it will change to True in 0.13. This would really unify the two functions. Thoughts?

Comment (dengemann) [inline mne/cov.py]: It should be False and not exposed. It makes only sense in some situations
and only when you have Epochs from several conditions.

Comment (Eric89GXL) [inline mne/cov.py]: > It should be False and not exposed.

So it's a bug how MNE-C has computed it in the `sample` dataset, too, then?

File: mne/tests/test_cov.py [hunk 5]
```diff
@@ -116,14 +116,31 @@ def test_cov_estimation_on_raw_segment():
...
+def test_cov_estimation_on_raw_reg():
+    """Test estimation from raw with regularization
+    """
+    raw = Raw(raw_fname, preload=True)
+    raw.info['sfreq'] /= 10.
+    raw = RawArray(raw._data[:, ::10].copy(), raw.info)  # decimate for speed
+    cov_mne = read_cov(erm_cov_fname)
+    with warnings.catch_warnings(record=True):  # too few samples
+        warnings.simplefilter('always')
+        # Don't use "shrunk" here, for some reason it makes Travis 2.7 hang...
```
Comment (jasmainak) [inline mne/tests/test_cov.py]: add an `XXX` ?

File: mne/tests/test_cov.py [hunk 6]
```diff
@@ -116,14 +116,31 @@ def test_cov_estimation_on_raw_segment():
...
         warnings.simplefilter('always')
         cov = compute_raw_covariance(raw_2)
-    assert_true(len(w) == 1)
+    assert_true(any('Too few samples' in str(ww.message) for ww in w))
+
+
+@slow_test
+@requires_sklearn_0_15
+def test_cov_estimation_on_raw_reg():
+    """Test estimation from raw with regularization
```
Comment (jasmainak) [inline mne/tests/test_cov.py]: single line docstrings should have the ending quotes in the same line. Sorry, it's a nitpick :)

Comment (Eric89GXL) [inline mne/tests/test_cov.py]: I'm pretty sure we have it both ways in the repo, have we agreed on that as the standard?

Comment (Eric89GXL) [inline mne/tests/test_cov.py]: (even this file seems to have more separate-liners than same-liners...)

Comment (jasmainak) [inline mne/tests/test_cov.py]: depends whether you want to follow PEP257 :)

**Comment by jasmainak** [comment]
code looks super-clean other than my nitpicks. I haven't seen the details, but I am sure you can discuss it with Denis and Alex.

File: mne/cov.py (hunk 4 above)
Comment (Eric89GXL) [inline mne/cov.py]: Actually I guess it's not from the `sample` dataset...

@agramfort @dengemann do you have any idea where the "erm cov" in our `mne/io/tests/data/test_erm-cov.fif` comes from? It doesn't match either from the `sample` dataset (according to a dumb test):
```
0fb0b2afb60ea4a6626b09ae142020bc MNE-sample-data/MEG/sample/ernoise-cov.fif
91cd4feea5bc8b18177351407d11787b MNE-sample-data/MEG/sample/sample_audvis-cov.fif
61a2d8033a3cabbed770adec6b2c5dde mne-python/mne/io/tests/data/test_erm-cov.fif
f8338b1cba1a8bf3a571ace3f72d97ff mne-python/mne/io/tests/data/test-cov.fif
```
Maybe we just computed it an incorrect way in the beginning and inherited it for `compute_covariance_raw`? So I guess changing the behavior to `keep_sample_mean=True` could be considered a bugfix?

File: mne/tests/test_cov.py (hunk 6 above)
Comment (Eric89GXL) [inline mne/tests/test_cov.py]: Arff, didn't know there was a PEP about this. We'll have to fix a lot of files to comply :)

Comment (agramfort) [inline mne/tests/test_cov.py]: ok let's fix when we see a mistake in a PR. Let's avoid a massive diff...

File: mne/cov.py (hunk 4 above)
Comment (agramfort) [inline mne/cov.py]: erm comes from the MGH repo of empty room recordings that are archived on a regular basis.

I think our compute_covariance_raw should match the default behavior of mne_process_raw

Comment (Eric89GXL) [inline mne/cov.py]: From the 2.7.3 manual:
```
If the recommended keepsamplemean option is specified in the covariance
matrix definition file, the baseline correction is applied to the epochs but
the means at individual samples are not subtracted.
```
So I think the problem is that the computation was done with the recommended option `keepsamplemean` turned off. In other words, the file in our repo is wrong, and our code has had a bug (bad default behavior) for a long time. I'll change it here and change `whats_new.py`. Thanks for slogging through this with me y'all.

**Comment by Eric89GXL** [comment]
@dengemann @agramfort I think I figured it out. It turns out that the way we do the computation in `master` is equivalent to using a single long interval, while throwing out any 0.2-sec (by default) chunks that violate the `reject` and `flat` criteria. I think this only coincidentally matched up with what `keep_sample_mean=False` provided before.

In any case, it means I had to do some tweaks in order to get the old behavior back. Take a look at the new docs and see if they make sense to you, and hopefully the tests show some minor differences.

**Comment by Eric89GXL** [comment]
I could add a kwarg `baseline_correct='raw' | 'epochs'` or something like that to expose the difference...? Then you could get equivalent behavior to `compute_covariance(Epochs(..., make_fixed_length_events(...), baseline=(None, None))` with `compute_raw_covariance`.

**Comment by Eric89GXL** [comment]
I exposed the `baseline` parameter to make it clear there are two ways of doing the baseline correction. Also rebased again.

**Comment by dengemann** [comment]
why can we do basleine correction at all here? With continous data this is a weird idea. It should not have such a parameter.

**Comment by Eric89GXL** [comment]
It's how the C code works and how the `master` version of this code works. Subtracting the mean of the entire data segment is what you'd do to compute the covariance normally, so it's logical to do it.

**Comment by Eric89GXL** [comment]
...or maybe you mean the `'epochs'` mode (baseline correct each segment) shouldn't be exposed?

**Comment by dengemann** [comment]
> ...or maybe you mean the 'epochs' mode (baseline correct each segment) shouldn't be exposed?

yep.

**Comment by Eric89GXL** [comment]
Ahh okay, I'll remove it and put the old explanation that was back in. I didn't hear back on the suggestion from last week so I thought maybe people thought it was reasonable.

**Comment by Eric89GXL** [comment]
Commit reverted, hopefully the explanation makes sense now.

**Comment by dengemann** [comment]
I still don't understand why we need a baseline parameter. In continous mode there is no baseline. Baseline to me is a concept from an event related context and means: the time of non-interest before t0.

**Comment by Eric89GXL** [comment]
There is no baseline parameter now

**Comment by dengemann** [comment]
Cool looks clean to me now.
We should at some point look what kind of results this code gives now across different conditions.

**Comment by Eric89GXL** [comment]
The only thing I wonder about, is does it matter for the advanced methods whether or not each epoch/segment it processes is actually zero mean? Because in this case it's zero-mean across the concatenation of all epochs, but each individual epoch will have a non-zero mean.

**Comment by dengemann** [comment]
> The only thing I wonder about, is does it matter for the advanced methods whether or not each epoch/segment it processes is actually zero mean? 

Not sure, we usually standard scale everything. I would have to look it up. Gut feeling it that it may matter if certain parameters are chosen.

**Comment by agramfort** [comment]
you need to rebase

**Comment by Eric89GXL** [comment]
Do you have time to look now? If so I'm happy to rebase. If not, I'd rather wait -- I think I've rebased a couple times already.

File: mne/cov.py [hunk 7]
```diff
@@ -386,66 +403,82 @@ def compute_raw_covariance(raw, tmin=None, tmax=None, tstep=0.2,
...
-    logger.info('[done]')
-
-    ch_names = [raw.info['ch_names'][k] for k in picks]
-    bads = [b for b in raw.info['bads'] if b in ch_names]
-    projs = cp.deepcopy(raw.info['projs'])
-    # XXX : do not compute eig and eigvec now (think it's better...)
-    return Covariance(data, ch_names, bads, projs, nfree=n_samples)
+    epochs = Epochs(raw, events, 1, 0, tstep_m1, baseline=None,
+                    picks=picks, reject=reject, flat=flat, verbose=False,
+                    preload=True, proj=False)
```
Comment (agramfort) [inline mne/cov.py]: does this induce an increased memory usage? what if you compute the noise cov on 5 mins of empty room? does it blow up the memory?

Comment (Eric89GXL) [inline mne/cov.py]: It will increase memory usage. For a 5 minute file this would be ~1GB:
```
5 * 60 * 1000 * 366 * 8
878400000
```
I don't see a way around it if we want to be able to use the advanced methods, though.

**Comment by Eric89GXL** [comment]
@agramfort any idea about whether or not each segment needs to be zero-mean (per channel) to emply the `sklearn` methods? The way we `compute_raw_covariance`, the channels are zero-mean across all time and epochs, whereas typically with `compute_covariance` the channels are zero-mean across time for each epoch individually.

**Comment by dengemann** [comment]
> whereas typically with compute_covariance the channels are zero-mean across time for each epoch individually.

but this is not true by default, actually we assume that the data is already zero mean. see `assume_centered` flags.

**Comment by Eric89GXL** [comment]
> but this is not true by default

Ahh, I had assumed that it *was* assumed to be true, because I was thinking about the use case where epochs were created with `baseline=(None, None)` (which would imply zero mean per channel per epoch). So it sounds like you're saying that the sklearn reg methods do not require this, which is great.

**Comment by dengemann** [comment]
> Ahh, I had assumed that it was assumed to be true, 

sorry I mean we assume that the data are zero mean and sklearn will not demean the data.

**Comment by dengemann** [comment]
practically you either filter at 1Hz or so or you baseline *before* computing the cov.

**Comment by Eric89GXL** [comment]
> we assume that the data are zero mean and sklearn will not demean the data.

Hmm, I think we can't guarantee this for `compute_raw_covariance`. Perhaps we should check for a highpass and only allow non-'empirical' mode if it's e.g. 0.5 or higher?

**Comment by dengemann** [comment]
> Perhaps we should check for a highpass and only allow non-'empirical' mode if it's e.g. 0.5 or higher?

This is pretty independent of the other modes, each of them would know how to deal with nonzero mean data. Typically the raw cov should have similar filter settings as your data (at least the manual has been suggesting this). Maybe just a warning. The idea is anyways that you should compute your cov for inverse solutions on roughly zero mean data, it may otherwise produce model violations.

File: mne/cov.py [hunk 8]
```diff
@@ -386,66 +403,82 @@ def compute_raw_covariance(raw, tmin=None, tmax=None, tstep=0.2,
...
+    # This makes it equivalent to what we used to do, treating all epochs
+    # as if they were a single long one...
+    ch_means = epochs._data.mean(axis=0).mean(axis=1)
+    epochs._data -= ch_means[np.newaxis, :, np.newaxis]
+    # fake this value so there are no complaints from compute_covariance...
+    epochs.baseline = (None, None)
+    return compute_covariance(epochs, keep_sample_mean=True, method=method,
```
Comment (dengemann) [inline mne/cov.py]: why is this True? keep_sample_mean is only meaningful if you have different conditions and it only works with 'empirical'. It is a legacy option hat is not very often used I assume.

Comment (Eric89GXL) [inline mne/cov.py]: It's the equivalent of what `compute_raw_covariance` was doing under the hood.

Comment (Eric89GXL) [inline mne/cov.py]: And actually `keep_sample_mean=True` is the default for `compute_covariance`. Maybe you're thinking of `keep_sample_mean=False` as being legacy?

Comment (dengemann) [inline mne/cov.py]: ... maybe we have a bug.

Comment (dengemann) [inline mne/cov.py]: No it's fine. Forget what I said. The API is a bit weird semantically.

Comment (dengemann) [inline mne/cov.py]: It means if you put it to False, it will be subtracted.

Comment (dengemann) [inline mne/cov.py]: We're fine.

Comment (dengemann) [inline mne/cov.py]: We're fine.

Comment (agramfort) [inline mne/cov.py]: @Eric89GXL what seems really weird to me is that if you remove the sample mean across epochs the output will depend on the tstep ... this tstep should only be there to reject really bad segments.

the option to remove mean across epochs should not be there for raw data.

Comment (Eric89GXL) [inline mne/cov.py]: Currently there is no option to remove the sample mean *for each epoch*, you can only remove the mean *across all epochs* which will not depend on `tstep`. So I think what you're saying is already accounted for.

File: mne/cov.py [hunk 9]
```diff
@@ -386,66 +403,82 @@ def compute_raw_covariance(raw, tmin=None, tmax=None, tstep=0.2,
...
-    bads = [b for b in raw.info['bads'] if b in ch_names]
-    projs = cp.deepcopy(raw.info['projs'])
-    # XXX : do not compute eig and eigvec now (think it's better...)
-    return Covariance(data, ch_names, bads, projs, nfree=n_samples)
+    epochs = Epochs(raw, events, 1, 0, tstep_m1, baseline=None,
+                    picks=picks, reject=reject, flat=flat, verbose=False,
+                    preload=True, proj=False)
+    # This makes it equivalent to what we used to do, treating all epochs
+    # as if they were a single long one...
+    ch_means = epochs._data.mean(axis=0).mean(axis=1)
```
Comment (agramfort) [inline mne/cov.py]: ok just saw it there...

File: mne/cov.py [hunk 10]
```diff
@@ -903,7 +947,7 @@ def _gaussian_loglik_scorer(est, X, y=None):
     n_samples, n_features = X.shape
     log_like = np.zeros(n_samples)
     log_like = -.5 * (X * (np.dot(X, precision))).sum(axis=1)
-    log_like -= .5 * (n_features * log(2. * np.pi) - _logdet(precision))
+    log_like -= .5 * (n_features * np.log(2. * np.pi) - _logdet(precision))
```
Comment (agramfort) [inline mne/cov.py]: why was this necessary? I prefer to use math module for operations on native types such as float

File: mne/cov.py [hunk 11]
```diff
@@ -345,28 +345,45 @@ def _check_n_samples(n_samples, n_chan):
...
 
-    Note: To speed up the computation you should consider preloading raw data
-    by setting preload=True when reading the Raw data.
+    .. note:: This function will:
+
+                  1. Partition the data into evenly spaced, equal-length
+                     epochs.
+                  2. Load them into memory.
+                  3. Subtract the mean across all time points and epochs
+                     for each channel.
```
Comment (Eric89GXL) [inline mne/cov.py]: @agramfort this is the corresponding doc that hopefully makes it clear.

File: mne/cov.py (hunk 10 above)
Comment (Eric89GXL) [inline mne/cov.py]: `np.log` is used elsewhere in the calculation so it saves an import and makes it consistent in use, but I'll revert it

**Comment by agramfort** [comment]
that's it for my nitpicks. Did not test it but I assume you're confident we're good? :)

I just want to avoid breaking something in such a core part of the MNE pipeline.

**Comment by Eric89GXL** [comment]
The old tests still pass and the calculations should be the same as before. The only drawback *should* be increased memory usage. So yeah I'm happy :)

**Comment by agramfort** [comment]
ok then rebase and make CIs happy then +1 for merge :)
